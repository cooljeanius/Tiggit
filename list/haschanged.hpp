#ifndef __HASCHANGED_HPP_
#define __HASCHANGED_HPP_

#include "listbase.hpp"

namespace List
{
  /* A simple callback class you can use to get notified when a list
     has changed. Just set the list you want to "monitor" as the
     parent.

     This class is so simple it really isn't necessary - it provides
     convenient class and member function names more than anything
     else.
   */
  class HasChanged : public ListBase
  {
  protected:
    HasChanged(ListBase *par)
      : ListBase(par) {}

    virtual void notify() = 0;

  private:
    void updateList() { notify(); }
  };
}
#endif
