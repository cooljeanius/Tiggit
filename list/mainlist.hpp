#ifndef __MAINLIST_HPP_
#define __MAINLIST_HPP_

#include "listbase.hpp"

namespace List
{
  /* A main list gets all its elements from an external list. It has
     no parent, but is used as parent for the other list classes.
   */
  class MainList : public ListBase
  {
  public:
    // Use this to set up the list
    PtrList &fillList() { return list; }

    // Then call this when the list is complete, if you want children
    // to be automatically updated.
    void done() { updateChildren(); }

  private:
    void updateList() {}
  };
}
#endif
