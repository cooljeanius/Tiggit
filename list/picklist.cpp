#include "picklist.hpp"
#include <assert.h>

using namespace List;

PickList::PickList(ListBase *par)
  : ListBase(par), pck(NULL)
{ assert(par); }

void PickList::setPick(Picker *p)
{
  pck = p;
  updateChildren();
}

void PickList::updateList()
{
  const PtrList &par = ((ListBase*)parent)->getList();

  list.resize(0);
  for(int i=0; i<par.size(); i++)
    if((pck == NULL) || pck->include(par[i]))
      list.push_back(par[i]);
}
