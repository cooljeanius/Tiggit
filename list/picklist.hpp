#ifndef __PICKLIST_HPP_
#define __PICKLIST_HPP_

#include "listbase.hpp"

namespace List
{
  struct Picker
  {
    virtual bool include(const void* x) = 0;
    virtual ~Picker() {}
  };

  /* A picklist is used to pick out elements from a parent list, based
     on an externally set criterium.
   */
  class PickList : public ListBase
  {
    Picker *pck;

  public:
    PickList(ListBase *par);

    // Set picker. A NULL value means 'pick everything'.
    void setPick(Picker *p = NULL);

  private:
    void updateList();
  };
}
#endif
