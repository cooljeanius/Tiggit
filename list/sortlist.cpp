#include <assert.h>
#include <algorithm>
#include "sortlist.hpp"

using namespace List;

SortList::SortList(ListBase *par)
  : ListBase(par), srt(NULL), reverse(false)
{ assert(par); }

void SortList::setSort(Sorter *p)
{
  srt = p;
  updateChildren();
}

void SortList::setReverse(bool b)
{
  reverse = b;
  updateChildren();
}

// Intermediary needed to work around std::sort() suckiness
struct Tmp
{
  Sorter *srt;

  bool operator()(const void *a, const void *b)
  { return srt->isLess(a,b); }
};

void SortList::updateList()
{
  const PtrList &par = ((ListBase*)parent)->getList();

  // First, copy the list over
  list = par;

  // If there is no sorter, do simplified version
  if(!srt)
    {
      if(reverse)
        std::reverse(list.begin(), list.end());

      return;
    }

  Tmp tmp;
  tmp.srt = srt;

  if(reverse)
    std::sort(list.rbegin(), list.rend(), tmp);
  else
    std::sort(list.begin(), list.end(), tmp);
}
