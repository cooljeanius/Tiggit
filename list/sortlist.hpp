#ifndef __SORTLIST_HPP_
#define __SORTLIST_HPP_

#include "listbase.hpp"

namespace List
{
  struct Sorter
  {
    // Return true if a should come before b
    virtual bool isLess(const void* a, const void* b) = 0;
    virtual ~Sorter() {}
  };

  /* A sortlist sorts the parent list based on the given sorter.
   */
  class SortList : public ListBase
  {
  public:
    SortList(ListBase *par);

    // Set sorter. A NULL value will copy the parent list unsorted
    // (but reverse status still applies.)
    void setSort(Sorter *p = NULL);
    void setReverse(bool b);
    bool getReverse() { return reverse; }

  private:
    Sorter *srt;
    bool reverse;

    void updateList();
  };
}
#endif
