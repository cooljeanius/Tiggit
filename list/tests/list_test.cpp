#include <iostream>
#include "picklist.hpp"
#include "sortlist.hpp"
#include "mainlist.hpp"
#include "haschanged.hpp"

using namespace std;
using namespace List;

int ii(const void*p) { return *((int*)p); }
void *vp(int i)
{
  int *ip = new int;
  *ip = i;
  return ip;
}

void print(const ListBase &l)
{
  const PtrList &lst = l.getList();
  for(int i=0; i<lst.size(); i++)
    cout << " " << ii(lst[i]);
  cout << endl;
}

void print(const std::string &msg, const ListBase &l)
{
  cout << msg << ":";
  print(l);
}

struct PrintChange : HasChanged
{
  std::string what;

  PrintChange(ListBase *l, const std::string &w)
    : HasChanged(l), what(w) {}

  void notify()
  {
    cout << what << " was updated\n";
  }
};

struct TestSort : Sorter
{
  bool isLess(const void *a, const void *b)
  {
    int A = ii(a), B = ii(b);

    if(A == 8 && B != 8) return true;
    if(B == 8) return false;

    if(A == 13 && B != 13) return false;
    if(B == 13) return true;

    return A < B;
  }
} tsort;

struct IntPicker : Picker
{
  bool include(const void *p) { return include(ii(p)); }
  virtual bool include(int i) = 0;
};

struct EvenPick : IntPicker
{
  bool include(int i)
  { return (i % 2) == 0; }
} even;

struct Pick5 : IntPicker
{
  bool include(int i)
  { return i > 5; }
} pick5;

int main()
{
  MainList ml;
  PickList pl(&ml);
  SortList sl(&pl);
  pl.setPick(&even);

  PrintChange p1(&ml, "Main list");
  PrintChange p2(&pl, "Pick list");
  PrintChange p3(&sl, "Sort list");

  cout << "Initial values:\n";
  print("Main", ml);
  print("Picker", pl);
  print("Sorter", sl);

  cout << "\nFilling in main list:\n";

  PtrList &lst = ml.fillList();
  for(int i=1;i<=10;i++)
    {
      lst.push_back(vp(i));
    }
  ml.done();
  print("Main", ml);
  print("Picker", pl);
  print("Sorter", sl);

  cout << "\nReversing sort:\n";
  sl.setReverse(true);
  print("Main", ml);
  print("Picker", pl);
  print("Sorter", sl);

  cout << "\nAdding some new elements:\n";
  lst.push_back(vp(111));
  lst.push_back(vp(200));
  lst[6] = vp(13);
  ml.done();
  print("Main", ml);
  print("Picker", pl);
  print("Sorter", sl);

  cout << "\nChanging sorter:\n";
  sl.setSort(&tsort);
  print("Main", ml);
  print("Picker", pl);
  print("Sorter", sl);

  cout << "\nChanging picker:\n";
  pl.setPick(&pick5);
  print("Main", ml);
  print("Picker", pl);
  print("Sorter", sl);

  cout << "\nReversing sort again:\n";
  sl.setReverse(true);
  print("Main", ml);
  print("Picker", pl);
  print("Sorter", sl);
  return 0;
}
