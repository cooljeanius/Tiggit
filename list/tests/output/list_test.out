Initial values:
Main:
Picker:
Sorter:

Filling in main list:
Sort list was updated
Pick list was updated
Main list was updated
Main: 1 2 3 4 5 6 7 8 9 10
Picker: 2 4 6 8 10
Sorter: 2 4 6 8 10

Reversing sort:
Sort list was updated
Main: 1 2 3 4 5 6 7 8 9 10
Picker: 2 4 6 8 10
Sorter: 10 8 6 4 2

Adding some new elements:
Sort list was updated
Pick list was updated
Main list was updated
Main: 1 2 3 4 5 6 13 8 9 10 111 200
Picker: 2 4 6 8 10 200
Sorter: 200 10 8 6 4 2

Changing sorter:
Sort list was updated
Main: 1 2 3 4 5 6 13 8 9 10 111 200
Picker: 2 4 6 8 10 200
Sorter: 8 2 4 6 10 200

Changing picker:
Sort list was updated
Pick list was updated
Main: 1 2 3 4 5 6 13 8 9 10 111 200
Picker: 6 13 8 9 10 111 200
Sorter: 8 6 9 10 111 200 13

Reversing sort again:
Sort list was updated
Main: 1 2 3 4 5 6 13 8 9 10 111 200
Picker: 6 13 8 9 10 111 200
Sorter: 13 200 111 10 9 6 8
